crates/proto/src/xfer/dnssec_dns_handle/mod.rs

```diff
@@ -790,23 +806,52 @@ where
     //  TODO: strip RRSIGS to accepted algorithms and make algorithms configurable.
     let verifications = rrsigs
         .iter()
-        .map(|rrsig| {
+        .enumerate()
+        .filter_map(|(i,rrsig)| {
             let handle = handle.clone_with_context();
             let query = Query::query(rrsig.data().signer_name().clone(), RecordType::DNSKEY);
 
+            if i > MAX_RRSIGS_PER_RRSET {
+                warn!("too many ({i}) RRSIGs for rrset {rrset:?}; skipping");
+                return None;
+            }
+
             // TODO: Should this sig.signer_name should be confirmed to be in the same zone as the rrsigs and rrset?
-            handle
+            Some(handle
                 .lookup(query.clone(), options)
                 .first_answer()
                 .map_err(|proto| {
                     ProofError::new(Proof::Indeterminate, ProofErrorKind::Proto { query, proto })
                 })
                 .map_ok(|message| {
+                    let mut tag_count = HashMap::<u16, usize>::new();
+
                     // DNSKEYs were already validated by the inner query in the above lookup
                     let dnskeys = message
                         .answers()
                         .iter()
-                        .filter_map(|r| r.try_borrow::<DNSKEY>());
+                        .filter_map(|r| {
+                            let dnskey = r.try_borrow::<DNSKEY>()?;
+
+                            match dnskey.data().calculate_key_tag() {
+                                Ok(tag) => match tag_count.get_mut(&tag) {
```

Comment:
I'd suggest yielding the `tag` outside of the `match` to reduce rightward drift, like:

```rust
                            let tag = match dnskey.data().calculate_key_tag() {
                                Ok(tag) => tag,
                                Err(e) => {
                                    warn!("unable to calculate key tag: {e:?}; skipping key");
                                    return None;
                                }
                            };

                            match tag_count.get_mut(&tag) {
                                Some(n_keys) => {
                                    *n_keys += 1;
                                    if *n_keys > MAX_KEY_TAG_COLLISIONS {
                                        warn!("too many ({n_keys}) DNSKEYs with key tag {tag}; skipping");
                                        return None;
                                    }
                               }
                               None => _ = tag_count.insert(tag, 1),
                            }
```